### ISSUE TITLE
Drop ordering requirement on successive Aggregate requests

### ISSUE THREAD

**Issue by BranLwyd** [abetterinternet/ppm-specification#217]
Currently, the successive aggregate requests in an aggregation job are required to include client report data (report shares/transitions) in a consistent order.

Example language from the spec:
> The sequence of Transition messages corresponds to the ReportShare sequence of the AggregateInitReq. The order of these sequences MUST be the same (i.e., the nonce of the first Transition MUST be the same as first ReportShare and so on).

> Next, the helper processes the Transition messages from the leader. If any message appears out of order or has an unrecognized nonce, or if any two messages have the same nonce, then the helper MUST abort with error "unrecognizedMessage".

The justification I have heard for this requirement is that requiring ordering saves implementations from needing to do an `O(n lg n)` sort, where `n` is the number of client reports being processed. However, even without ordering, no sort or other `O(n lg n)` operation would be required from implementations. Specifically, the aggregator can construct a hashmap from report nonce to report share/transition (requiring `O(n)` time total); then, when matching report shares/transitions to stored state, do an `O(1)` map lookup per report, for a total of `O(n)` time spent doing lookups.

An upside to dropping the ordering requirement is a simpler specification which permits a simpler implementation with fewer error modes.

A downside to dropping the ordering requirement is that, even though both ordered and non-ordered implementations can do matching of request parts to stored state in `O(n)` time, I suspect that the unordered implementation will take more wallclock time (e.g. have "larger constants hidden by the big-O") due to the hashmap operations. [Not measured yet.]

Questions:
* Is there a justification other than avoiding an `O(n lg n)` sort to requiring consistency of ordering in aggregate requests?

**Comment by BranLwyd** [comment]
(this is a consideration post-interop-pilot, most likely--unless all parties think a change is a good idea, I think the interop pilot keeping the ordering requirement is fine)

**Comment by cjpatton** [comment]
You're right that O(n) is not hard to achieve. However, not all implementors necessarily have a geneirc HashMap structure (or even a generic quicksort algorithm) on hand. For example, imagine you were implementing this in C and couldn't use stdlib. The naive algorithm for matching a request to a response would take O(n^2) time.

The current design follows the following design principle: Try to avoid designing the protocol in a way that permits bad (i.e., inefficient) implementations. In particular, the goal of the current spec is to only allow one "natural" matching algorithm. Perhaps it would be helpful to spell this algorithm out.

FWIW, this exact problem came up in the design of Encrypted Client Hello. (See https://github.com/tlswg/draft-ietf-tls-esni/issues/378.) There the motivation was to reduce the risk of bad implementations becoming DoS vectors. Here the risk is even more pronounced, since the number of reports processed in a single run of the aggregation sub-protocol might be 100s or even 1000s.

**Comment by cjpatton** [comment]
Here's how we ended up solving the problem in ECH: https://github.com/davidben/draft-ietf-tls-esni/commit/f8e9df6c475f9268b3f0b9054c14f76ca9cbe9d7#

**Comment by BranLwyd** [comment]
Closing this for now -- this is a minor request, and I think the arguments for keeping the ordering requirement are good. We can reopen if this question comes up again.
